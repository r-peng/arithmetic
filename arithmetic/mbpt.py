import numpy as np
import networkx as nx
import quimb.tensor as qtn
class HD:
    def __init__(self,idx,sym):
        self.idx = idx
        self.n = len(idx) // 2
        self.sym = sym
        self.pix = self.n + 1 # dummy

        b1 = []
        for i in range(self.n):
            li,ri = idx[2*i],idx[2*i+1]
            if li==i:
                b1.append(i)
            if ri==i:
                b1.append(i)
        self.b1 = b1
class HDprop:
    def __init__(self,idx,sign,pix):
        self.idx = idx
        self.n = len(idx) // 2
        self.sign = sign
        ne = 0
        for i in range(self.n):
            li,ri = idx[2*i], idx[2*i+1]
            if li==ri:
                ne += 1
        self.ne = ne
        self.pix = pix

        G = nx.MultiDiGraph()
        G.add_nodes_from(range(self.n),op=False) # add all hugenholtz vertices
        G.nodes[pix]['op'] = True
        for i in range(self.n):
            li,ri = idx[2*i],idx[2*i+1]
            G.add_edge(i,li)
            G.add_edge(i,ri)
        self.G = G
def load_diagrams(n,fname='/home/rppeng/code/EDHD/Examples/HugenDiag'):
    # n=order
    out = open(fname+f'{n}.txt','r').readlines()
    for l in out:
        if l[:4]=='Diag':
            str_ = l.split('=')[-1]
            while not str_[0].isnumeric():
                str_ = str_[1:]
            while not str_[-1].isnumeric():
                str_ = str_[:-1] 
            diags = str_.split('},{')
            for ix in range(len(diags)):
                diag = diags[ix].split(',')
                diag = tuple([int(c)-1 for c in diag])
                diags[ix] = diag
        if l[:3]=='Sym': 
            str_ = l.split('=')[-1]
            while not str_[0].isnumeric():
                str_ = str_[1:]
            while not str_[-1].isnumeric():
                str_ = str_[:-1] 
            syms = str_.split(',')
            syms = [int(c) for c in syms]
    return [HD(diags[ix],syms[ix]) for ix in range(len(diags))]
def distribute_diagrams(diags):
    hfc_map = {i:[] for i in range(diags[0].n+1)}
    while len(diags)!=0:
        diag = diags.pop()
        hfc_map[len(diag.b1)].append(diag)
    return hfc_map
def generate_property(diag):
    nhfc = len(diag.b1)
    if nhfc==0:
        return []
    idx = diag.idx
    sign = 1 if diag.sym>0 else -1
    ls = [HDprop(idx,sign,pix) for pix in diag.b1]
    if nhfc==1:
        return ls[0]
    rix = set()
    nm = nx.algorithms.isomorphism.categorical_node_match('op',False)
    for i in range(nhfc):
        for j in range(i+1,nhfc):
            if nx.is_isomorphic(ls[i].G,ls[j].G,node_match=nm):
                rix.add(j)
    lix = list(set(range(nhfc)).difference(rix))
    return [ls[i] for i in lix]
def theta(t):
    return (1.+np.sign(t))/2.
def get_propagator(mo_energies,occ_vec,t1,t2):
    norb = len(mo_energies)
    ng1,ng2 = len(t1),len(t2)
    g = np.zeros((norb,ng1,ng2,2,2),dtype=complex) #t1,t2,a1,a2
    for p in range(norb):
        for i1 in range(ng1):
            for i2 in range(ng2):
                t12 = t1[i1]-t2[i2]
                phase = np.exp(-1j*mo_energies[p]*t12)
                gpm = -1j*phase * (1-occ_vec[p])
                gmp = 1j*phase * occ_vec[p]
                g[p,i1,i2,1,0] = gpm 
                g[p,i1,i2,0,1] = gmp
                g[p,i1,i2,0,0] = theta(t12)*gpm + theta(-t12)*gmp
                g[p,i1,i2,1,1] = theta(-t12)*gpm + theta(t12)*gmp
    symm0 = np.linalg.norm(g[...,0,0]+g[...,1,1]-g[...,0,1]-g[...,1,0])
    symm1 = np.linalg.norm(g[...,0,0]+g[...,1,1].transpose(0,2,1).conj())
    symm2 = np.linalg.norm(g[...,0,1]+g[...,0,1].transpose(0,2,1).conj())
    symm3 = np.linalg.norm(g[...,1,0]+g[...,1,0].transpose(0,2,1).conj())
    if symm0 + symm1 + symm2 + symm3 > 1e-10:
        print(symm0,symm1,symm2,symm3)
    return g
def get_tn(diag,mo_energies,occ_vec,v2,to,ti,w):
    # to: open times
    # ti: integrated times
    gii = get_propagator(mo_energies,occ_vec,ti,ti)
    gio = get_propagator(mo_energies,occ_vec,ti,to)
    goi = get_propagator(mo_energies,occ_vec,to,ti)
    tn = qtn.TensorNetwork([])
    n,pix,idx = diag.n,diag.pix,diag.idx
    def select(oix,iix):
        if oix==pix:
            return 1j*goi
        elif iix==pix:
            return 1j*gio
        else:
            return 1j*gii
    for i in range(n):
        li,ri = idx[2*i],idx[2*i+1]
        if i == pix:
            oix = li if ri==i else ri
            tag = 'p' if ri==i else 'q'
            tn.add_tensor(qtn.Tensor(data=select(oix,i),
                                     inds=(f'{tag}{oix}',f't{oix}',f't{i}',f'a{oix}',f'a{i}')))
            p_,p = f'{tag}{oix}',f'{tag}{i}'
        else:
            tn.add_tensor(qtn.Tensor(data=select(li,i),
                                     inds=(f'p{li}',f't{li}',f't{i}',f'a{li}',f'a{i}')))
            tn.add_tensor(qtn.Tensor(data=select(ri,i),
                                     inds=(f'q{ri}',f't{ri}',f't{i}',f'a{ri}',f'a{i}')))
            tn.add_tensor(qtn.Tensor(data=1j*v2.copy(),
                                     inds=(f'p{li}',f'q{ri}',f'p{i}',f'q{i}')))
    tid1,tid2 = tn.ind_map[f'a{pix}']
    g1,g2 = tn.tensor_map[tid1],tn.tensor_map[tid2]
    if p in g1.inds and p_ in g2.inds:
        g1.reindex_({f'a{pix}':'a',f't{pix}':'t'})
        g2.reindex_({f'a{pix}':'a_',f't{pix}':'t_'})
    elif p in g2.inds and p_ in g1.inds:
        g2.reindex_({f'a{pix}':'a',f't{pix}':'t'})
        g1.reindex_({f'a{pix}':'a_',f't{pix}':'t_'})
    else:
        print(g1)
        print(g2)
        print(p,p_)
        exit()
    for i in range(n):
        if i!=pix:
            tn.add_tensor(qtn.Tensor(data=w.copy(),inds=(f't{i}',)))
            tn.add_tensor(qtn.Tensor(data=np.array([1.,-1.]),inds=(f'a{i}',)))
    return tn,(p,'t','a',p_,'t_','a_')
def _get_tn(diag,mo_energies=None,occ_vec=None,v2=None,v1=None,t=None,w=None):
    norb = len(mo_energies)
    ng = len(t)
    g = np.zeros((norb,ng,ng,2,2),dtype=complex)
    for p in range(norb):
        for i1 in range(ng):
            for i2 in range(ng):
                phase = np.exp(-1j*mo_energies[p]*(t[i1]-t[i2]))
                g[p,i1,i2,1,0] = -1j*phase * (1-occ_vec[p])
                g[p,i1,i2,0,1] = 1j*phase * occ_vec[p]
                if i1>i2:
                    g[p,i1,i2,0,0] = g[p,i1,i2,1,0] 
                    g[p,i1,i2,1,1] = g[p,i1,i2,0,1] 
                elif i2>i1:
                    g[p,i1,i2,0,0] = g[p,i1,i2,0,1]
                    g[p,i1,i2,1,1] = g[p,i1,i2,1,0]
                #else:
                #    g[p,i1,i2,0,0] = g[p,i1,i2,1,1] = (g[p,i1,i2,0,1]+g[p,i1,i2,1,0])/2.
    #print(np.linalg.norm(g[...,0,0]+g[...,1,1]-g[...,0,1]-g[...,1,0]))
    print(np.linalg.norm(g[...,0,0]+g[...,1,1].transpose(0,2,1).conj()))
    print(np.linalg.norm(g[...,0,1]+g[...,0,1].transpose(0,2,1).conj()))
    print(np.linalg.norm(g[...,1,0]+g[...,1,0].transpose(0,2,1).conj()))

    tn = qtn.TensorNetwork([])
    n,pix,idx = diag.n,diag.pix,diag.idx
    for i in range(n):
        li,ri = idx[2*i],idx[2*i+1]
        if li==i:
            tn.add_tensor(qtn.Tensor(data=g.copy(),
                                     inds=(f'q{ri}',f't{ri}',f't{i}',f'a{ri}',f'a{i}')))
            if i == pix:
                output_inds = f'q{ri}',f'q{i}'
                #tn.add_tensor(qtn.Tensor(data=np.array([1.,-1.]),
                #                     inds=(f'a{i}',)))
            else:
                tn.add_tensor(qtn.Tensor(data=v1.copy(),
                                     inds=(f'q{ri}',f'q{i}')))
        elif ri==i:
            tn.add_tensor(qtn.Tensor(data=g.copy(),
                                     inds=(f'p{li}',f't{li}',f't{i}',f'a{li}',f'a{i}')))
            if i == pix:
                output_inds = f'p{li}',f'p{i}'
                #tn.add_tensor(qtn.Tensor(data=np.array([1.,-1.]),
                #                     inds=(f'a{i}',)))
            else:
                tn.add_tensor(qtn.Tensor(data=-1j*v1.copy(),
                                     inds=(f'p{li}',f'p{i}')))
        else:
            tn.add_tensor(qtn.Tensor(data=g.copy(),
                                     inds=(f'p{li}',f't{li}',f't{i}',f'a{li}',f'a{i}')))
            tn.add_tensor(qtn.Tensor(data=g.copy(),
                                     inds=(f'q{ri}',f't{ri}',f't{i}',f'a{ri}',f'a{i}')))
            tn.add_tensor(qtn.Tensor(data=-1j*v2.copy(),
                                     inds=(f'p{li}',f'q{ri}',f'p{i}',f'q{i}')))
    
    output_inds = output_inds + tuple([qtn.rand_uuid() for i in range(2)])
    tid1,tid2 = tn.ind_map[f'a{pix}']
    g1,g2 = tn.tensor_map[tid1],tn.tensor_map[tid2]
    if output_inds[0] in g1.inds:
        assert output_inds[1] in g2.inds
        g1.reindex_({f'a{pix}':output_inds[2]})
        g2.reindex_({f'a{pix}':output_inds[3]})
    elif output_inds[0] in g2.inds:
        assert output_inds[1] in g1.inds
        g2.reindex_({f'a{pix}':output_inds[2]})
        g1.reindex_({f'a{pix}':output_inds[3]})
    else:
        print(g1)
        print(g2)
        print(output_inds)
        exit()
    for i in range(n):
        if i!=pix:
            tn.add_tensor(qtn.Tensor(data=w.copy(),inds=(f't{pix}',f't{i}')))
            tn.add_tensor(qtn.Tensor(data=np.array([1.,-1.]),inds=(f'a{i}',)))
    return tn,(f't{pix}',)+output_inds
