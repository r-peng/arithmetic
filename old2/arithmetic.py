import numpy as np
import math, operator
from functools import reduce
np.set_printoptions(suppress=True)
def _svd(A,naxis,thresh=1e-12,bdim=20): # svd after the 1st naxis
    tdim = A.shape
    mdim1 = reduce(operator.mul,tdim[:naxis]) 
    mdim2 = reduce(operator.mul,tdim[naxis:])
    u,s,vh = np.linalg.svd(A.reshape(mdim1,mdim2),full_matrices=False)
#    print(tdim,s)
#    print('svd truncation err: ', s[bdim])
    s = s[s>thresh] if s[0]>thresh else s[:1]
    s = s[:bdim] if len(s)>bdim else s
    u = np.reshape(u[:,:len(s)],tdim[:naxis]+(len(s),))
    vh = np.reshape(vh[:len(s),:],(len(s),)+tdim[naxis:])
    return u,s,vh
def _contract(tn,ins): 
    tmp = [np.einsum('axb,x->ab',tn[i],ins[i]) for i in range(len(ins))]
    if len(tmp)==1:
        return tmp[0]
    else: 
        return np.linalg.multi_dot(tmp)
def _get_bdim(tn):
    return [tn[i].shape[-1] for i in range(len(tn))]
def _left_sweep(tn,thresh=1e-12,bdim=20):
    out = []
    vh = tn[0].copy()
    for i in range(1,len(tn)):
        A = np.einsum('axb,byd->axyd',vh,tn[i])
        u,s,vh = _svd(A,2,thresh,bdim)
        s = np.sqrt(s)
        u = np.einsum('...a,a->...a',u,s)
        vh = np.einsum('a...,a->a...',vh,s)
        out.append(u)
    out.append(vh)
    assert len(out)==len(tn)
    return out
def _right_sweep(tn,thresh=1e-12,bdim=20):
    out = []
    u = tn[-1].copy()
    for i in range(len(tn)-2,-1,-1):
        A = np.einsum('axb,byd->axyd',tn[i],u)
        u,s,vh = _svd(A,2,thresh,bdim)
        s = np.sqrt(s)
        u = np.einsum('...a,a->...a',u,s)
        vh = np.einsum('a...,a->a...',vh,s)
        out.insert(0,vh)
    out.insert(0,u)
    assert len(out)==len(tn)
    return out
def _compare_list(l1,l2):
    if len(l1)!=len(l2):
        return False
    else:
        out = True
        for i in range(len(l1)):
            if l1[i]!=l2[i]:
                out = False
                break
        return out
def _iterate(tn,thresh=1e-12,bdim=20,max_iter=50):
    old = tn.copy()
    err = 1.0
    for i in range(max_iter):
        tmp = _left_sweep(old,thresh,bdim)
        new = _right_sweep(tmp,thresh,bdim)
        if _compare_list(_get_bdim(old),_get_bdim(new)):
#            err = sum([np.linalg.norm(old[i]-new[i]) for i in range(len(tn))])
#        if err < 1e-6:
            break 
        old = new.copy()
    return old
def _multiply_const(tn,a,i=None,thresh=1e-12,bdim=20,max_iter=50):
    if i is None:
        if a > thresh:
            a = a**(1/len(tn))
            out = [x*a for x in tn]
        elif a < -thresh:
            a = (-a)**(1/len(tn))
            out = [x*a for x in tn]
            out[0] *= -1
        else:
            out = [np.zeros_like(x) for x in tn]
    else: 
        out = tn.copy()
        out[i] = out[i]*a
#    out = _iterate(out,thresh,bdim,max_iter)
    return out
def _multiply_concat(tn1,tn2,thresh=1e-12,bdim=20,max_iter=50):
    # tn1: x1...xk
    # tn2: y1...yl
    # return x1...xk,y1...yl
    out = tn1 + tn2
    out = _iterate(out,thresh,bdim,max_iter)
    return out
def _multiply(tn1,tn2,thresh=1e-12,bdim=20,max_iter=50):
    assert len(tn1)==len(tn2)
    assert tn2[0].shape[0] == 1
    vh  = np.einsum('axb,xc->axbc',tn1[0],tn2[0][0,:,:])
    out = []
    for i in range(1,len(tn1)):
        A = np.einsum('axbc,byd,cye->axyde',vh,tn1[i],tn2[i])
        u,s,vh = _svd(A,2,thresh,bdim)
        s = np.sqrt(s)
        u = np.einsum('...a,a->...a',u,s)
        vh = np.einsum('a...,a->a...',vh,s)
        out.append(u)
    b,y,d,e = vh.shape
    assert d==1
    assert e==1
    vh = vh.reshape((b,y,d))
    out.append(vh)
    assert len(out)==len(tn1)
    out = _iterate(out,thresh,bdim,max_iter)
    return out
def _const(d,a,i=None,thresh=1e-12,bdim=20,max_iter=50):
    if i is None:
        if a > thresh:
            a = a**(1/len(d))
            out = [np.ones((1,di,1))*a for di in d]
        elif a < -thresh:
            a = (-a)**(1/len(d))
            out = [np.ones((1,di,1))*a for di in d]
            out[0] *= -1
        else:
            out = [np.zeros((1,di,1))*a for di in d]
    else:
        out = [np.ones((1,d[i],1)) for i in range(len(d))]
        out[i] = out[i]*a
#    out = _iterate(out,thresh,bdim,max_iter)
    return out
def _add_const(tn,b,i=None,thresh=1e-12,bdim=20,max_iter=50):
    d = [tn[i].shape[1] for i in range(len(tn))]
    b = _const(d,b,i,thresh,bdim,max_iter)
    return _add(tn,b,thresh,bdim,max_iter)
def _add_concat(tn1,tn2,thresh=1e-12,bdim=20,max_iter=50):
    # tn1: x1...xk
    # tn2: y1...yl
    # return x1...xk,y1...yl
    dx = [tn1[i].shape[1] for i in range(len(tn1))]
    bx = _const(dx,1.0,0,thresh,bdim,max_iter)
    dy = [tn2[i].shape[1] for i in range(len(tn2))]
    by = _const(dy,1.0,0,thresh,bdim,max_iter)
    f = tn1+by
    g = bx+tn2
    return _add(f,g,thresh,bdim,max_iter)
def _add(tn1,tn2,thresh=1e-12,bdim=20,max_iter=50):
    assert len(tn1)==len(tn2)
    for i in range(len(tn1)):
        assert tn1[i].shape[1]==tn2[i].shape[1]
    if len(tn1)==1:
        return [tn1[0]+tn2[0]]
    else: 
        out = []
        a1,x,b1 = tn1[0].shape
        a2,x,b2 = tn2[0].shape
        assert a1==1
        assert a2==1
        vh = np.zeros((1,x,b1+b2))
        vh[0,:,:b1] = tn1[0][0,:,:].copy()
        vh[0,:,b1:] = tn2[0][0,:,:].copy()
        for i in range(1,len(tn1)-1):
            a1,x,b1 = tn1[i].shape
            a2,x,b2 = tn2[i].shape
            tmp = np.zeros((a1+a2,x,b1+b2))
            tmp[:a1,:,:b1] = tn1[i]
            tmp[a1:,:,b1:] = tn2[i]
            A = np.einsum('axb,byc->axyc',vh,tmp)
            u,s,vh = _svd(A,2,thresh,bdim) 
            s = np.sqrt(s)
            u = np.einsum('...a,a->...a',u,s)
            vh = np.einsum('a...,a->a...',vh,s)
            out.append(u)
        a1,x,b1 = tn1[-1].shape
        a2,x,b2 = tn2[-1].shape
        assert b1==1
        assert b2==1
        tmp = np.zeros((a1+a2,x,1))
        tmp[:a1,:,0] = tn1[-1][:,:,0].copy()
        tmp[a1:,:,0] = tn2[-1][:,:,0].copy()
        A = np.einsum('axb,byc->axyc',vh,tmp)
        u,s,vh = _svd(A,2,thresh,bdim) 
        s = np.sqrt(s)
        u = np.einsum('...a,a->...a',u,s)
        vh = np.einsum('a...,a->a...',vh,s)
        out = out+[u,vh]
        assert len(out)==len(tn1)
        out = _iterate(out,thresh,bdim,max_iter)
        return out
def _poly1(tn,a,thresh=1e-12,bdim=20,max_iter=50):
    # a = [a0,...,ap]
    powers = [None,tn.copy()]
    for i in range(2,len(a)):
        powers.append(_multiply(tn,powers[-1],thresh,bdim,max_iter))
#        print('power={},bdim={}'.format(i,_get_bdim(powers[-1])))
    assert len(powers)==len(a)
    term = _multiply_const(powers[1],a[1],None,thresh,bdim,max_iter)
    f = _add_const(term,a[0],None,thresh,bdim,max_iter)
    for i in range(2,len(a)):
        term = _multiply_const(powers[i],a[i],None,thresh,bdim,max_iter)
        f = _add(f,term,thresh,bdim,max_iter)
#        print('order={},bdim={}'.format(i,_get_bdim(f)))
    return f
def _poly2(tn,coeff,thresh=1e-12,bdim=20,max_iter=50):
    # Horner's method
    a = coeff.copy()
    a.reverse()
    f = _multiply_const(tn,a[0],None,thresh,bdim,max_iter)
    f = _add_const(f,a[1],None,thresh,bdim,max_iter)
    for i in range(2,len(a)):
        f = _multiply(f,tn,thresh,bdim,max_iter)
        f = _add_const(f,a[i],None,thresh,bdim,max_iter)
#        print('power={},bdim={}'.format(i,_get_bdim(f)))
    return f
if __name__=='__main__':
    import math
    p = 10
    a = [1.0/math.factorial(i) for i in range(p+1)]
    
    n = 10 
    d = 10
    thresh = 1e-12
    bdim = 20
    max_iter = 50
    x = np.random.rand(n,d)/n
    i = np.zeros(d)
    i[0] = 1
    ins = [i for j in range(n)]
    tn = [x[0,:].reshape(1,d,1)]
    for i in range(1,n):
        tn = _add_concat(tn,[x[i,:].reshape(1,d,1)],thresh,bdim,max_iter)
    out = _contract(tn,ins)
    print('err',abs(out[0,0]-sum(x[:,0])),_get_bdim(tn))
    #exit()
    f1 = _poly1(tn,a,thresh,bdim,max_iter)
    f2 = _poly2(tn,a,thresh,bdim,max_iter)
    out1 = _contract(f1,ins)
    out2 = _contract(f2,ins)
    print('err1',abs(out1[0,0]-np.exp(sum(x[:,0]))))
    print('err2',abs(out2[0,0]-np.exp(sum(x[:,0]))))
