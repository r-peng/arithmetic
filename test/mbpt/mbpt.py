import numpy as np
from arithmetic.mbpt import (
    load_diagrams,
    distribute_diagrams,
    generate_property,
    get_tn,
)
np.set_printoptions(precision=6,suppress=True,linewidth=1000)
norb = 6
na = 2
nb = na
occ_vec = np.array([1]*na + [0]*(norb-na))
occ_vec_ = np.concatenate([occ_vec,occ_vec])

t_leads = 1.
t_hyb = .4
U = 1.
idot = norb // 2

h1 = np.zeros((norb,norb))
for i in range(norb):
    if i < norb - 1:
        dot = (i == idot or i + 1 == idot)
        h1[i, i + 1] = -t_hyb if dot else -t_leads
    if i > 0:
        dot = (i == idot or i - 1 == idot)
        h1[i, i - 1] = -t_hyb if dot else -t_leads
e,v = np.linalg.eigh(h1)
print('E0=',sum(e[:na])*2)
e_ = np.concatenate([e,e])
print(e_)
print(occ_vec_)

eri = np.zeros((norb,)*4)
eri[idot,idot,idot,idot] = U
eri = np.einsum('ijkl,ip,jq,kr,ls->pqrs',eri,v,v,v,v)
eri_aa = eri - eri.transpose(1,0,2,3)
eri_ab = eri.copy()
#print(np.linalg.norm(eri_aa+eri_aa.transpose(1,0,2,3)))
#print(np.linalg.norm(eri_ab-eri_ab.transpose(1,0,3,2)))
#ov = np.einsum('ijcb,cbaj->ia',eri_aa[:na,:na,na:,na:],eri_aa[na:,na:,na:,:na])
#ov += 2*np.einsum('ijcb,cbaj->ia',eri_ab[:na,:na,na:,na:],eri_ab[na:,na:,na:,:na])
#ov -= np.einsum('kjab,ibkj->ia',eri_aa[:na,:na,na:,na:],eri_aa[:na,na:,:na,:na])
#ov -= 2*np.einsum('kjab,ibkj->ia',eri_ab[:na,:na,na:,na:],eri_ab[:na,na:,:na,:na])
#vv = np.einsum('ikac,bcik->ba',eri_aa[:na,:na,na:,na:],eri_aa[na:,na:,:na,:na])
#vv += 2*np.einsum('ikac,bcik->ba',eri_ab[:na,:na,na:,na:],eri_ab[na:,na:,:na,:na])
#oo = -np.einsum('ikac,acjk->ij',eri_aa[:na,:na,na:,na:],eri_aa[na:,na:,:na,:na])
#oo -= 2*np.einsum('ikac,acjk->ij',eri_ab[:na,:na,na:,na:],eri_ab[na:,na:,:na,:na])
#rdm1 = np.zeros((norb,norb))
#rdm1[:na,:na] = oo
#rdm1[na:,na:] = vv
#rdm1[:na,na:] = ov
#rdm1[na:,:na] = ov.T
#print(rdm1)
eri_ = np.zeros((norb*2,)*4)
eri_[:norb,:norb,:norb,:norb] = eri_aa.copy()
eri_[norb:,norb:,norb:,norb:] = eri_aa.copy()
eri_[:norb,norb:,:norb,norb:] = eri_ab.copy()
eri_[norb:,:norb,norb:,:norb] = eri_ab.copy()
eri_[:norb,norb:,norb:,:norb] = -eri_ab.transpose(1,0,2,3).copy()
eri_[norb:,:norb,:norb,norb:] = -eri_ab.transpose(1,0,2,3).copy()
print('check eri symmetry=',np.linalg.norm(eri_+eri_.transpose(1,0,2,3)))

tf = 1.
dt = .1
t = np.arange(0,tf,dt)
ng = len(t)
w = np.ones(ng)*dt

rdm1_ = np.zeros(norb,dtype=complex)
rdm1_[:na] = 1.
rdm1_ = np.diag(rdm1_)
rdm1_ = np.stack([rdm1_]*ng,axis=0)
for n in [3,4]:
    diags = load_diagrams(n)
    hfc_map = distribute_diagrams(diags)
    print('##################### '+f'order={n-1},ndiag={len(hfc_map[1])}'+' #####################')
    for ix,diag_ in enumerate(hfc_map[1]):
        diag = generate_property(diag_)
        print(ix,diag.idx)
        tn,output_inds = get_tn(diag,e_,occ_vec_,eri_,to=t,ti=t,w=w)
        tn.exponent -= sum([np.log10(float(i)) for i in range(2,n)])
        tn.exponent -= np.log10(2.) * diag.ne
        print(tn)
        #print(output_inds)
        #exit()
        #tn.full_simplify_(seq='ADCRSL',output_inds=output_inds,equalize_norms=1.,progbar=True)
        #print(tn)
        ig = tn.contract(output_inds=output_inds).data
        aabb = np.linalg.norm(ig[:norb,:,:,:norb,:,:]-ig[norb:,:,:,norb:,:,:])
        ab = np.linalg.norm(ig[:norb,:,:,norb:,:,:])
        ba = np.linalg.norm(ig[norb:,:,:,:norb,:,:])
        print('check ab=',aabb,ab,ba)
        g = -1j*ig[:norb,:,:,:norb,:,:]
        symm0 = np.linalg.norm(g[:,:,0,:,:,0]+g[:,:,1,:,:,1]-g[:,:,0,:,:,1]-g[:,:,1,:,:,0])
        symm1 = np.linalg.norm(g[:,:,0,:,:,0]+g[:,:,1,:,:,1].transpose(2,3,0,1).conj())
        symm2 = np.linalg.norm(g[:,:,0,:,:,1]+g[:,:,0,:,:,1].transpose(2,3,0,1).conj())
        symm3 = np.linalg.norm(g[:,:,1,:,:,0]+g[:,:,1,:,:,0].transpose(2,3,0,1).conj())
        print('check g symm=',symm0,symm1,symm2,symm3)
        rdm1 = -1j*np.einsum('piqi->ipq',g[:,:,0,:,:,1])
        print('check rdm1 symm=',np.linalg.norm(rdm1-rdm1.transpose(0,2,1).conj()))
        rdm1_ += rdm1 * 10**tn.exponent * diag.sign
tr = np.einsum('nii->',rdm1_[...])
for i in range(ng):
    print(i,rdm1_[i,idot,idot],tr[i])
