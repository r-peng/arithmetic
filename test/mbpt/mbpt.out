E0= -2.9120439557122078
[-1.088566 -0.367456  0.367456  1.088566 -1.088566 -0.367456  0.367456  1.088566]
[1 1 0 0 1 1 0 0]
4.47213595499958
[[[ 0.      +0.j       -0.994081+0.108642j -0.976394+0.215997j -0.947148+0.320796j -0.90669 +0.421797j]
  [ 0.      +0.j        0.      +0.j       -0.994081+0.108642j -0.976394+0.215997j -0.947148+0.320796j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j       -0.994081+0.108642j -0.976394+0.215997j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j       -0.994081+0.108642j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]]

 [[ 0.      +0.j       -0.999325+0.036737j -0.997301+0.073425j -0.99393 +0.110014j -0.989218+0.146454j]
  [ 0.      +0.j        0.      +0.j       -0.999325+0.036737j -0.997301+0.073425j -0.99393 +0.110014j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j       -0.999325+0.036737j -0.997301+0.073425j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j       -0.999325+0.036737j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]]

 [[ 1.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.999325-0.036737j  1.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.997301-0.073425j  0.999325-0.036737j  1.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.99393 -0.110014j  0.997301-0.073425j  0.999325-0.036737j  1.      +0.j        0.      +0.j      ]
  [ 0.989218-0.146454j  0.99393 -0.110014j  0.997301-0.073425j  0.999325-0.036737j  1.      +0.j      ]]

 [[ 1.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.994081-0.108642j  1.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.976394-0.215997j  0.994081-0.108642j  1.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.947148-0.320796j  0.976394-0.215997j  0.994081-0.108642j  1.      +0.j        0.      +0.j      ]
  [ 0.90669 -0.421797j  0.947148-0.320796j  0.976394-0.215997j  0.994081-0.108642j  1.      +0.j      ]]]
[[[-1.      +0.j       -0.994081+0.108642j -0.976394+0.215997j -0.947148+0.320796j -0.90669 +0.421797j]
  [ 0.      +0.j       -1.      +0.j       -0.994081+0.108642j -0.976394+0.215997j -0.947148+0.320796j]
  [ 0.      +0.j        0.      +0.j       -1.      +0.j       -0.994081+0.108642j -0.976394+0.215997j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j       -1.      +0.j       -0.994081+0.108642j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j       -1.      +0.j      ]]

 [[-1.      +0.j       -0.999325+0.036737j -0.997301+0.073425j -0.99393 +0.110014j -0.989218+0.146454j]
  [ 0.      +0.j       -1.      +0.j       -0.999325+0.036737j -0.997301+0.073425j -0.99393 +0.110014j]
  [ 0.      +0.j        0.      +0.j       -1.      +0.j       -0.999325+0.036737j -0.997301+0.073425j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j       -1.      +0.j       -0.999325+0.036737j]
  [ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j       -1.      +0.j      ]]

 [[ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.999325-0.036737j  0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.997301-0.073425j  0.999325-0.036737j  0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.99393 -0.110014j  0.997301-0.073425j  0.999325-0.036737j  0.      +0.j        0.      +0.j      ]
  [ 0.989218-0.146454j  0.99393 -0.110014j  0.997301-0.073425j  0.999325-0.036737j  0.      +0.j      ]]

 [[ 0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.994081-0.108642j  0.      +0.j        0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.976394-0.215997j  0.994081-0.108642j  0.      +0.j        0.      +0.j        0.      +0.j      ]
  [ 0.947148-0.320796j  0.976394-0.215997j  0.994081-0.108642j  0.      +0.j        0.      +0.j      ]
  [ 0.90669 -0.421797j  0.947148-0.320796j  0.976394-0.215997j  0.994081-0.108642j  0.      +0.j      ]]]
